acme/api/api.go

```
@@ -0,0 +1,151 @@
+	c.Challenges = (*ChallengeService)(&c.common)
+	c.Orders = (*OrderService)(&c.common)
+
+	return c, nil
+}
+
+// post performs an HTTP POST request and parses the response body as JSON,
+// into the provided respBody object.
+func (a *Core) post(uri string, reqBody, response interface{}) (*http.Response, error) {
+	content, err := json.Marshal(reqBody)
+	if err != nil {
+		return nil, errors.New("failed to marshal message")
+	}
+
+	return a.retrievablePost(uri, content, response, 0)
+}
+
+// postAsGet performs an HTTP POST ("POST-as-GET") request.
+// https://tools.ietf.org/html/draft-ietf-acme-acme-16#section-6.3
+func (a *Core) postAsGet(uri string, response interface{}) (*http.Response, error) {
+	return a.retrievablePost(uri, []byte{}, response, 0)
+}
+
+func (a *Core) retrievablePost(uri string, content []byte, response interface{}, retry int) (*http.Response, error) {
+	resp, err := a.signedPost(uri, content, response)
+	if err != nil {
+		// during tests, 5 retries allow to support ~50% of bad nonce.
+		if retry >= 5 {
+			log.Infof("too many retry on a nonce error, retry count: %d", retry)
+			return resp, err
+		}
+		switch err.(type) {
+		// Retry once if the nonce was invalidated
+		case *acme.NonceError:
+			log.Infof("nonce error retry: %s", err)
+			resp, err = a.retrievablePost(uri, content, response, retry+1)
+			if err != nil {
+				return resp, err
+			}
+		default:
+			return resp, err
+		}
+	}
+
+	return resp, nil
+}
+
+func (a *Core) signedPost(uri string, content []byte, response interface{}) (*http.Response, error) {
+	signedContent, err := a.jws.SignContent(uri, content)
+	if err != nil {
+		return nil, fmt.Errorf("failed to post JWS message -> failed to sign content -> %v", err)
+	}
+
+	signedBody := bytes.NewBuffer([]byte(signedContent.FullSerialize()))
+
+	resp, err := a.do.Post(uri, signedBody, "application/jose+json", response)
+
+	nonce, nonceErr := nonces.GetFromResponse(resp)
+	if nonceErr == nil {
+		a.nonceManager.Push(nonce)
```

It seems like the error is discarded here, I always like to add a comment when this is the case.